cpp/src/community/dendrogram.cuh

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <memory>
+#include <rmm/device_buffer.hpp>
+#include <vector>
+
+namespace cugraph {
+
+template <typename vertex_t>
+class Dendrogram {
+ public:
+  Dendrogram() : level_size_(), level_ptr_() {}
+
+  void add_level(vertex_t num_verts)
+  {
+    cudaStream_t stream{0};
+    rmm::mr::device_memory_resource *mr = rmm::mr::get_current_device_resource();
+
+    level_ptr_.push_back(
+      std::make_unique<rmm::device_buffer>(num_verts * sizeof(vertex_t), stream, mr));
+    level_size_.push_back(num_verts);
+  }
+
+  size_t current_level() const { return level_size_.size() - 1; }
+
+  size_t num_levels() const { return level_size_.size(); }
+
+  vertex_t *get_level_ptr_unsafe(size_t level) const
```

Comment:
It is dangerous to have a `const` getter return a non-const pointer to a member. Something like this, for example, can happen:
```
template<typename T>
void foo(Dendrogram<T> const& d)
{
  d.get_level_ptr_unsafe(0)[0] = T{3};
}

int main(void)
{
  Dendrogram<int> d;

  foo(d);

  return 0;
}
```
`foo()` should not be permitted to modify the `const&` argument, `d`. But it can. This breaks the contract that the getter is supposed to fulfill (namely being an immutable getter).
(to compile this simple example, I changed `device_buffer` to `std::vector<vertex_t>`, but the idea is the same).

At the very least, `get_level_ptr_unsafe()` should not be a `const` member.

Reply:
Sorry, that was careless.  Next push will address this.

There should be a const version that returns a const pointer and a non-const version that returns a non-const pointer as it is used in both contexts.